### ISSUE TITLE
Composable Matchers - Negative Form

### ISSUE THREAD

**Issue by yelled3** [rspec/rspec-expectations#493]
PR: https://github.com/rspec/rspec-expectations/pull/505

see end of comment:

``` ruby
module RSpec
  module Matchers
    module Composable
      # Creates a compound `and` expectation. The matcher will
      # only pass if both sub-matchers pass.
      # This can be chained together to form an arbitrarily long
      # chain of matchers.
      #
      # @example
      #   expect(alphabet).to start_with("a").and end_with("z")
      #
      # @note The negative form (`expect(...).not_to matcher.and other`)
      #   is not supported at this time.
```

@myronmarston what are the plans for this?
seems reasonably straight forward...

**Comment by stevecj** [comment]
`define_negated_matcher` feels awkward to me as a solution for this issue.

One solution that I don't think has been suggested is do define `and_not_to` and `or_not_to` methods. I'm less sure about `or_not_to` since that might be hard to reason about when reading specs.

Anyway, `and_not_to` would allow us to write things like

    expect( foo ).
      not_to be_even.
      and_not_to be_between(5, 15)

**Comment by myronmarston** [comment]
@stevecj: At this point, `define_negated_matcher` shipped nearly 2 years ago and has seen plenty of usage in the wild, so we can't really consider removing it.  Your suggestion also doesn't really help when it comes to supporting composition...for example, with negated matchers you can do:

``` ruby
RSpec::Matchers.define_negated_matcher :an_array_excluding, :an_array_including

# ...

expect(result).to match [an_array_including(:foo), an_array_excluding(:foo)]
```

So I'd say that the approach that we went with is more composable and powerful than adding methods that would support negation in compound expectations but nowhere else.

If you want experiment with the API you've suggested I'd encourage you to create an extension gem that adds it.

**Comment by stevecj** [comment]
I feel like `define_negated_matcher` is very useful in any case, but `and_not` is a much simpler solution to cases like the example that I gave.

In any case, I like your suggestion of writing an extension gem. I believe I will take your advice and do that. 😄

**Comment by myronmarston** [comment]
@Lukom RSpec has a feature that provides something like what you want that also avoids the trap @pirj refers to: define negated forms of the matchers you want to use.  In your `spec_helper.rb` (or a similar file), add this:

``` ruby
RSpec::Matchers.define_negated_matcher :make_no_db_query_matching, :make_db_query
```

Then you can use the negated matcher like so:

``` ruby
expect { subject }
  .to make_no_db_query_matching(/something 1/)
  .and make_no_db_query_matching(/something 2/)
```

That'll do what you want and provide nicer (and less confusing) failure output than we ever could have provided with `and_not`.   Note that the name of the negated matcher is entirely up to you; I thought `make_no_db_query_matching` reads nicely given the argument (but would probably better be paired with `make_db_query_matching(...)`) but you could just call it `make_no_db_query` or `avoid_making_db_query` or whatever you want.  RSpec will take care of producing a failure message (and matcher description) for the negated matcher that is based on the phrasing you use for the matcher name.

**Comment by Lukom** [comment]
@pirj Sorry, your point is not convincing, all programming languages have tools for negation, in which way RSpec is different and shouldn't have it?

@myronmarston Imagine that you have 100 matchers, and you have to write ~garbage~ 100 custom negation forms of them, and then also remember all negation forms, was it `not_make_db_query` or `avoid_make...` or `avoid_making` or `make_no...`?

currently when I have

```ruby
expect { subject }.not_to make_db_query(/something/)
```
and matcher is triggered I have this error message:
```
expected #<Proc:0x00007f8afd5f73e8 ...:172> not to make db query /something/
```
It reads perfectly well

When I define negation matcher `not_make_db_query` and have this:
```
expect { subject }.to not_make_db_query(/something/)
```
I get this:
```
expected #<Proc:0x00007f8ae1d879a8 ...:172> to not make db query /something/
```
which reads a bit worse, but still ok. We could have `avoid_making_db_query` matcher and have `... to avoid making db query` which reads better, agree, though it's less intuitive – you have to not just add `not_` to matcher name but also slightly change the verb, also you have to agree on the new convention with the team

When I have this:
```ruby
expect { subject }.
  to not_make_db_query(/not_triggered/).
    and not_make_db_query(/something/)
```
I get:
```
expected #<Proc:0x00007f8ae754f000 ...:151> to not make db query /something/
```
Basically the same message as previous, it doesn't matter if matcher is after `to` or `and`

I imagine that if `and_not` is implemented we'd have the same perfect message as in the first case

**Comment by Lukom** [comment]
Makes sense. I like solution with `dont` – you don't need parentheses and don't need custom code 🙌 why is it not yet implemented in RSpec? 😅

Still I think `and_not` may be ok if implement a warning whenever there is complex condition inside.
